OpenMP required by `libperf.c`'s `ucx_perf_do_warmup`

Most code in UCX that uses OpenMP is guarded to check that OpenMP is enabled. If it is not, there is fallback logic that doesn't use OpenMP

For example this happens in `libperf_thread.c` where an OpenMP block is provided if OpenMP is found

https://github.com/openucx/ucx/blob/3c6b20db9fec8204cfe9f2744aa519efef9de8f8/src/tools/perf/lib/libperf_thread.c#L21-L25

Otherwise a non-OpenMP fallback is provided

https://github.com/openucx/ucx/blob/3c6b20db9fec8204cfe9f2744aa519efef9de8f8/src/tools/perf/lib/libperf_thread.c#L138-L147

As a result any `#pragma omp` lines live in the OpenMP enabled block

https://github.com/openucx/ucx/blob/3c6b20db9fec8204cfe9f2744aa519efef9de8f8/src/tools/perf/lib/libperf_thread.c#L44

So are not encountered outside it

<hr>

With `libperf.c` however the OpenMP `#include` is optional

https://github.com/openucx/ucx/blob/3c6b20db9fec8204cfe9f2744aa519efef9de8f8/src/tools/perf/lib/libperf.c#L26-L28

Though later in the code `#pragma omp` is not guarded

https://github.com/openucx/ucx/blob/3c6b20db9fec8204cfe9f2744aa519efef9de8f8/src/tools/perf/lib/libperf.c#L1702-L1706

As a result building without OpenMP leads to errors like this

```C
make[3]: Entering directory '/home/conda/feedstock_root/build_artifacts/ucx-split_1714181624237/work/src/tools/perf/lib'
  CC       libucxperf_la-libperf.lo
  CC       libucxperf_la-libperf_memory.lo
libperf.c: In function 'ucx_perf_do_warmup':
libperf.c:1356: error: ignoring '#pragma omp barrier' [-Werror=unknown-pragmas]
 1356 | #pragma omp barrier
      | 
libperf.c:1357: error: ignoring '#pragma omp single' [-Werror=unknown-pragmas]
 1357 | #pragma omp single copyprivate(status)
      | 
libperf.c:1360: error: ignoring '#pragma omp barrier' [-Werror=unknown-pragmas]
 1360 | #pragma omp barrier
      | 
```

This comes from a UCX 1.16.0 build. Here is [the full build log]( https://github.com/openucx/ucx/files/15136778/ucx-split-feedstock_20240426.14_linux_64_c_compiler_version11cuda_compilernvcccuda_compiler_version11.8cxx_compiler_version11.txt )

<hr>

Would it be possible to guard the OpenMP code above and provide a fallback to allow building without OpenMP?

--- a/buildlib/tools/builds.sh
+++ b/buildlib/tools/builds.sh
@@ -303,6 +303,10 @@ build_no_devx() {
 	build_gcc --with-devx=no
 }
 
+build_no_openmp() {
+	build_gcc --disable-openmp
+}
+
 build_gcc_debug_opt() {
 	build_gcc CFLAGS=-Og CXXFLAGS=-Og
 }
@@ -408,47 +412,45 @@ build_fuse() {
 	fi
 }
 
-#
-# Do a given task and update progress indicator
-#
-do_task() {
-	amount=$1
-	shift
-	# cleanup build dir before the task
-	[ -n "${ucx_build_dir}" ] && rm -rf ${ucx_build_dir}/*
-
-	$@
-
-	echo "##vso[task.setprogress value=$PROGRESS;]Progress Indicator"
-	PROGRESS=$((PROGRESS+amount))
-}
-
-
 az_init_modules
 prepare_build
 
-[ "${long_test}" = "yes" ] && prog=5 || prog=12
-
-do_task "${prog}" build_docs
-do_task "${prog}" build_debug
-do_task "${prog}" build_prof
-do_task "${prog}" build_ugni
-do_task "${prog}" build_cuda
-do_task "${prog}" build_rocm
-do_task "${prog}" build_no_verbs
-do_task "${prog}" build_release_pkg
-do_task "${prog}" build_cmake_examples
-do_task "${prog}" build_fuse
-
+tests=('build_docs' \
+		'build_debug' \
+		'build_prof' \
+		'build_ugni' \
+		'build_cuda' \
+		'build_rocm' \
+		'build_no_verbs' \
+		'build_release_pkg' \
+		'build_cmake_examples' \
+		'build_fuse')
 if [ "${long_test}" = "yes" ]
 then
-	do_task 5 check_config_h
-	do_task 5 check_inst_headers
-	do_task 10 build_icc
-	do_task 10 build_pgi
-	do_task 10 build_gcc
-	do_task 8 build_no_devx
-	do_task 10 build_gcc_debug_opt
-	do_task 10 build_clang
-	do_task 10 build_armclang
+	tests+=('check_config_h' \
+			'check_inst_headers' \
+			'build_icc'\
+			'build_pgi' \
+			'build_gcc' \
+			'build_no_devx' \
+			'build_no_openmp' \
+			'build_gcc_debug_opt' \
+			'build_clang' \
+			'build_armclang')
 fi
+
+num_tests=${#tests[@]}
+for ((i=0;i<${num_tests};++i))
+do
+	test_name=${tests[$i]}
+
+	# update progress indicator
+	progress=$(( (i * 100) / num_tests ))
+	echo "##vso[task.setprogress value=${progress};]${test_name}"
+
+	# cleanup build dir before the task
+	[ -d "${ucx_build_dir}" ] && rm -rf ${ucx_build_dir}/*
+
+	# run the test
+	${test_name}
+done

--- a/src/tools/perf/lib/libperf.c
+++ b/src/tools/perf/lib/libperf.c
@@ -1696,16 +1696,17 @@ ucx_perf_do_warmup(ucx_perf_context_t *perf, const ucx_perf_params_t *params)
         /* Stop when reaching the deadline */
         stop_status = (ucs_get_time() > deadline) ? UCS_OK : UCS_INPROGRESS;
 
-        if (params->thread_count == 1) {
+        /* Synchronize on whether to continue or stop the warmup phase */
+#if _OPENMP
+#  pragma omp barrier
+#  pragma omp single copyprivate(status)
+#endif
+        {
             status = ucx_perf_test_exchange_status(perf, stop_status);
-        } else {
-#pragma omp barrier
-#pragma omp single copyprivate(status)
-            /* Synchronize on whether to continue or stop the warmup phase */
-            status = ucx_perf_test_exchange_status(perf, stop_status);
-#pragma omp barrier
         }
-
+#if _OPENMP
+#  pragma omp barrier
+#endif
         if (status != UCS_INPROGRESS) {
             return status;
         }

--- a/src/tools/perf/mad/perftest_mad.c
+++ b/src/tools/perf/mad/perftest_mad.c
@@ -316,8 +316,10 @@ static ucs_status_t perftest_mad_barrier(perftest_mad_rte_group_t *group)
     ucs_status_t status = UCS_OK;
     unsigned value;
 
-#pragma omp barrier
-#pragma omp single copyprivate(status)
+#if _OPENMP
+#  pragma omp barrier
+#  pragma omp single copyprivate(status)
+#endif
     {
         value = ++mad_magic;
 
@@ -331,7 +333,9 @@ static ucs_status_t perftest_mad_barrier(perftest_mad_rte_group_t *group)
 
         status = perftest_mad_recv_magic(group, ~value);
     }
-#pragma omp barrier
+#if _OPENMP
+#  pragma omp barrier
+#endif
     return status;
 }
 

--- a/src/tools/perf/perftest.c
+++ b/src/tools/perf/perftest.c
@@ -228,9 +228,10 @@ static unsigned sock_rte_group_index(void *rte_group)
 static void sock_rte_barrier(void *rte_group, void (*progress)(void *arg),
                              void *arg)
 {
-#pragma omp barrier
-
-#pragma omp master
+#if _OPENMP
+#  pragma omp barrier
+#  pragma omp master
+#endif
   {
     sock_rte_group_t *group = rte_group;
 
@@ -251,7 +252,10 @@ static void sock_rte_barrier(void *rte_group, void (*progress)(void *arg),
         }
     }
   }
-#pragma omp barrier
+
+#if _OPENMP
+#  pragma omp barrier
+#endif
 }
 
 static void sock_rte_post_vec(void *rte_group, const struct iovec *iovec,

--- a/test/apps/test_init_mt.c
+++ b/test/apps/test_init_mt.c
@@ -28,7 +28,9 @@ int main(int argc, char **argv)
     printf("starting test [%ld.%06ld] .. ", start.tv_sec, start.tv_usec);
     fflush(stdout);
 
-#pragma omp parallel
+#if _OPENMP
+#  pragma omp parallel
+#endif
     {
         ucs_status_t ctx_status, worker_status;
         ucp_context_h context;
@@ -47,7 +49,9 @@ int main(int argc, char **argv)
             }
         }
 
-#pragma omp barrier
+#if _OPENMP
+#  pragma omp barrier
+#endif
 
         if (ctx_status == UCS_OK) {
             if (worker_status == UCS_OK) {
@@ -57,7 +61,9 @@ int main(int argc, char **argv)
         }
     }
 
-#pragma omp barrier
+#if _OPENMP
+#  pragma omp barrier
+#endif
 
     gettimeofday(&finish, NULL);
     printf("[%ld.%06ld] finished %d threads\n",

